File: github_code/rod-cutting/java/52.java
package dynamicprogramming.rodcutting;

/*
 * Given a rod of length n, 
 * a table of prices pi for i=1,2,...n (ie. pi is the price of rod of length i)
 * determine the max revenue rn, obtainable by cutting up the rod and selling the pieces.
 * NOTE: if the price of pn of rod of length n is large enough, an optimal solution may require no cutting at all.
 * 
 * length i | 1  2  3  4  5   6   7   8   9   10
 * ---------|----------------------------------------------------------------------------------
 * price pi | 1  5  8  9  10  17  17  20  24  30
 * 
 */
public class RodCutting {
	
	int [] prices;
	int n ;
	
	int [] ri;
	
	public RodCutting(){}
	public RodCutting(int[] p, int n){
		this.prices = p;
		this.n = n;
		
		ri = new int[p.length];
	}
	
	/*
	 * Top down approach.
	 * while calculating the top optimum cost, all the bottom optimum are calculated and memoized
	 */
	public int memoizedCutRodAux(int[] p, int n, int[] r){
		int q=0;
		
		if(r[n] >= 0)	return r[n];
		
		if(n==0){
			q = 0;
		}else{
			q = -9999;
			for(int i=1; i<=n; i++){
				q = Math.max(q, p[i] + memoizedCutRodAux(p, n-i, r));
			}
		}
		r[n] = q;
		
		System.out.println("returning revenue for n=" + n + " is " + q);
		return q;
	}
	
	/*
	 * Bottom top approach
	 */
	public int bottomUpRod(int[] p, int n, int[] r){
		int q =-9999;
		
		r[0] = 0;
		for(int i=1; i<=n; i++){
			q = -9999;
			if(i==11){
				q = -9999;
			}
			for(int j=1; j<=i; j++){
				try{
					
					q = Math.max(q, p[j] + r[i-j]);
				}catch(RuntimeException e){
					System.out.println(" i=" + i + ", j=" + j);
					e.printStackTrace();
				}
			}
			r[i] = q;
		}
		return r[n];
	}
	
	
	public static void main(String[] args) {
		int [] pi = {0,1,5,8,9,10,17,17,20,24,30};// prices
		int n = 9;
		int[] r = new int[n+1];
		
		for(int i=0; i<=n; i++) r[i] = -9999;
		
		//to support n > pi.length
		if(n>pi.length){
			int[] pi2 = new int[n+1];
			for(int i=0; i<pi.length; i++) pi2[i] = pi[i];
			for(int i=pi.length; i<n; i++) pi2[i] = pi[pi.length-1];
			pi = pi2;
		}
		
		RodCutting rc = new RodCutting(pi, n);
		int optimumRevenue ;
		//optimumRevenue= rc.memoizedCutRodAux(pi, n, r);
		optimumRevenue = rc.bottomUpRod(pi, n, r);
		
		System.out.println("optimal revenue=" + optimumRevenue);
	}
}

/*
 * top down approach: take a length l, and find an optimal solution for n-l
 * buttom up approach: go building optimal solution from the buttom
 */

File: github_code/rod-cutting/cpp/52.c
/*
The rod-cutting problem is the following. Given a rod of length n inches and a
table of prices p i for i D 1; 2; : : : ; n, determine the maximum revenue r n obtain-
able by cutting up the rod and selling the pieces. Note that if the price p n for a rod
of length n is large enough, an optimal solution may require no cutting at all.
*/
#include <iostream>
#include "../../cpplib/cpputils.h"
using namespace std;

// O(exponential) // return the optimal price recursively, assumed prices[0] = 0 
// in this method cuts are used just to create the final answer of optimal cuts.
// algorithm does not require or uses cuts array
int findCuttingRecNonOptimal(int* prices, const int& priceLength, int rodSize, int* cuts)
{
	cout << "findCuttingRecNonOptimal called with : rodSize = " << rodSize << endl; 
	if( 0 == rodSize )
		return 0;
	
	int price = -1;
	// find the first cut and recurse on rest of the rodLength
	for( int i = 1 ; i <= rodSize ; i++ )
	{
		cout << "trying for the rod cut size of : " << i << endl;
		int nextPrice = prices[i] + findCuttingRecNonOptimal(prices, priceLength,rodSize-i,cuts);
		if( price < nextPrice )
		{
			cuts[rodSize] = i ; 
			price = nextPrice;
		}
	}
	cout << "returning final price for rodSize = " << rodSize << ", price " << price << endl;
	return price;
}

// O(n2) // return the optimal price recursively, assumed prices[0] = 0
// This method is top to bottom dynamic programming approach.
// in this method cuts are used just to create the final answer of optimal cuts.
// algorithm does not require or uses cuts array
int findCuttingRecOptimal(int* prices, const int& priceLength, int rodSize, int* cuts, int* optimalPrices)
{
	cout << "findCuttingRecOptimal called with : rodSize = " << rodSize << endl; 
	if( 0 == rodSize )
		return 0;
	
	int price = optimalPrices[rodSize]; // quick lookup if the price is already present.
	if( price != -1 )
		return price ;
	// find the first cut and recurse on rest of the rodLength
	for( int i = 1 ; i <= rodSize ; i++ )
	{
		cout << "trying for the rod cut size of : " << i << endl;
		int nextPrice = prices[i] + findCuttingRecOptimal(prices, priceLength,rodSize-i,cuts,optimalPrices);
		if( price < nextPrice )
		{
			cuts[rodSize] = i ; 
			price = nextPrice;
		}
	}
	cout << "returning final price for rodSize = " << rodSize << ", price " << price << endl;
	optimalPrices[rodSize] = price; // set the optimal price of the rodSize's cutting for quick lookup.
	return price;
}

/*
O(n2)
this is bottom up dynamic programming approach
*/
int findCuttingIterOptimal(int* prices, const int& priceLength, int rodSize, int* cuts, int* optimalPrices)
{
	optimalPrices[0] = 0 ;
	optimalPrices[1] = prices[1]; // optimal cut size of rod of length 1 is the price of rodSize = 1
	for(int i = 1 ; i <= rodSize; i++ ) // finding optimal price for rodSize = i
	{
		int maxPrice = -1;
		for( int j = 1 ; j <= i ; j++ ) // finding optimal first cut at j for rodSize = i 
		{
			int price = prices[j] + optimalPrices[i-j];
			if( maxPrice < price )
			{
				cuts[i] = j ;
				maxPrice = price;
			}
		}
		optimalPrices[i] = maxPrice; // set the optimal price of this rodSize for future use.
	}

	return optimalPrices[rodSize];
}

int main()
{
	int size = 0 ; 
	cout << "Input number of prices : " ;
	cin >> size;
	int* prices;
	cout << "Input the prices : " ;
	inputAndCreateArray(prices,size);

	print1DArray("The prices you entered : ", prices, size);

	cout << "Enter the size of the rod : ";
	int rodSize;
	cin >> rodSize;

	int finalPrice;
	int* cuts ; // what is optimal first cut for the rodSize == index of the array
	int* optimalPrices; // it stores the optimal prices for the different rodSizes for optimal algos
	createAndInitArray(cuts,rodSize+1,0);
	createAndInitArray(optimalPrices,rodSize+1,-1);	

	// finalPrice = findCuttingRecNonOptimal(prices, size, rodSize,cuts);
	// finalPrice = findCuttingRecOptimal(prices,size,rodSize,cuts,optimalPrices);
	finalPrice = findCuttingIterOptimal(prices,size,rodSize,cuts,optimalPrices);

	print1DArray("The optimal cuts are : ", cuts, rodSize+1);
	cout << "And the Optimal Price is : " << finalPrice << endl;

	delete cuts;
	delete prices;

}